src/autosave.js

```diff
@@ -0,0 +1,176 @@
+ *						return saveEditorsContentToDatabase( data );
+ *					}
+ *				};
+ *			} );
+ */
+export default class Autosave extends Plugin {
+	static get pluginName() {
+		return 'Autosave';
+	}
+
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * @member {Provider}
+		 */
+		this.provider = undefined;
+
+		/**
+		 * @protected
+		 * @type {Function}
+		 */
+		this._throttledSave = throttle( this._save.bind( this ), 500 );
+
+		/**
+		 * @private
+		 * @type {DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DomEmitterMixin );
+
+		/**
+		 * @protected
+		 * @type {Number}
+		 */
+		this._lastDocumentVersion = editor.model.document.version;
+	}
+
+	init() {
+		const editor = this.editor;
+		const doc = editor.model.document;
+		const pendingActions = editor.plugins.get( PendingActions );
+
+		this.listenTo( doc, 'change', this._saveIfDocumentChanged.bind( this ) );
+
+		// Flush on the editor's destroy listener with the highest priority to ensure that
+		// `editor.getData()` will be called before plugins are destroyed.
+		this.listenTo( editor, 'destroy', () => this._save(), { priority: 'highest' } );
+
+		// It's not possible to easy test it because karma uses `beforeunload` event
+		// to warn before full page reload and this event cannot be dispatched manually.
+		/* istanbul ignore next */
+		this._domEmitter.listenTo( window, 'beforeunload', ( evtInfo, domEvt ) => {
+			if ( pendingActions.isPending ) {
+				domEvt.returnValue = pendingActions.first.message;
+			}
+		} );
+	}
+
+	destroy() {
+		this._throttledSave.cancel();
+		this._domEmitter.stopListening();
+		super.destroy();
+	}
+
+	/**
+	 * @protected
+	 */
+	_flush() {
+		this._throttledSave.flush();
+	}
+
+	/**
+	 * Filters out changes in document, that don't impact on the data (e.g. selection changes).
+	 * Quickly returns when finds an operation that potentially changes document's data.
+	 *
+	 * @private
+	 */
+	_saveIfDocumentChanged( evt, batch ) {
+		for ( const delta of batch.deltas ) {
+			for ( const operation of delta.operations ) {
+				const name = operation.name;
+
+				if ( !name || !isUserSelectionOperation( operation ) ) {
+					this._throttledSave();
+
+					return;
+				}
+			}
+		}
+	}
+
+	/**
+	 * If the provider is set and new document version exists,
+	 * `_save()` method creates a pending action and calls `provider.save()` method.
+	 * It waits for the result and then removes the created pending action.
+	 *
+	 * @private
+	 */
+	_save() {
+		if ( !this.provider ) {
+			return;
+		}
+
+		const version = this.editor.model.document.version;
+
+		if ( version <= this._lastDocumentVersion ) {
+			return;
+		}
+
+		this._lastDocumentVersion = version;
+
+		const pendingActions = this.editor.plugins.get( PendingActions );
+		const action = pendingActions.add( 'Saving in progress.' );
+
+		Promise.resolve( this.provider.save() )
+			.then( () => {
+				pendingActions.remove( action );
```

Comment:
Pending action should not be removed if new changes happen in the meantime. These should be a test about it.

Reply:
We'd have to track all promises to be able to implement your idea. With the current one, you get new pending action per each change (with the fix of above issue), so there'll be no moment when there's some saving action and you don't have any pending action. But few pending action with `'Saving in progress.' ` may happen at some points.

Reply:
In all other places where we add pending actions (comments, upload) we ensure that only one pending action is added by a plugin. One can get all pending actions and will be wrong if he will have multiple 'Saving in progress.' actions. Instead, we can have a collection of promises to check if there were changes.